smarts/core/utils/logging.py

```diff
@@ -29,3 +34,50 @@ def timeit(name: str):
     ellapsed_time = (time() - start) * 1000
 
     logging.info(f'"{name}" took: {ellapsed_time:4f}ms')
+
+
+libc = ctypes.CDLL(None)
+try:
+    c_stdout = ctypes.c_void_p.in_dll(libc, "stdout")
+except:
+    # macOS
+    c_stdout = ctypes.c_void_p.in_dll(libc, "__stdoutp")
+
+
+def try_fsync(fd):
+    try:
+        os.fsync(fd)
+    except OSError:
+        # On GH actions was returning an OSError: [Errno 22] Invalid argument
+        pass
+
+
+@contextmanager
+def surpress_stdout():
+    original_stdout_fno = sys.stdout.fileno()
+
+    # XXX: Range is to prevent collisions if there are race conditions with multiple
+    #      processes calling redirect_stdout.
+    dup_stdout_fno = randrange(5, 128)
+    os.dup2(original_stdout_fno, dup_stdout_fno)
+
+    devnull_fno = os.open(os.devnull, os.O_WRONLY)
+    os.dup2(devnull_fno, original_stdout_fno)
+    sys.stdout = os.fdopen(devnull_fno, "w")
+
+    try:
+        yield
+    finally:
+        sys.stdout.flush()
+        libc.fflush(c_stdout)
+        try_fsync(devnull_fno)
+        os.close(devnull_fno)
+
+        os.dup2(dup_stdout_fno, original_stdout_fno)
+        try_fsync(dup_stdout_fno)
+        sys.stdout = os.fdopen(dup_stdout_fno, "w")
+
+        def close():
+            sys.stdout.close()
+
+        atexit.register(close)
```

Comment:
It looks like this is registered every time `supress_stdout()` is used? I have seen this as a leak through global objects before.

Reply:
That's a good point, let me think of a solution.

Reply:
This is intentional, notice it's closing the dup stream which is unique per process and changes per run.

Reply:
More ideally, we could reuse the old stdout (1), but I was unable to. Can tinker w/ it in the future. That would allow removing the need for the atexit.register.

Reply:
Ok, I just improved the implementation. Now we don't need the close.

Reply:
OK, please mark it somehow as being a problem. It will leak not so nicely since it is being called to wrap the `traci.connect()` code which can get called an uncapped number times when running SMARTS.